src/Domain/DomainHelpers.cpp

```diff
@@ -0,0 +1,369 @@
+  for (size_t i = 0; i < common.size(); i++) {
+    Point<VolumeDim, Frame::Grid> mapped_pt1 = map_of_block1(pts1[i]);
+    Point<VolumeDim, Frame::Grid> mapped_pt2 = map_of_block2(pts2[i]);
+    for (size_t j = 0; j < VolumeDim; j++) {
+      summed_squared_distances += square(mapped_pt1[j] - mapped_pt2[j]);
+    }
+  }
+  return sqrt(summed_squared_distances / two_to_the(VolumeDim - 1));
+}
+
+template <size_t VolumeDim>
+size_t DomainHelpers<VolumeDim>::find_block_id(
+    const std::array<size_t, two_to_the(VolumeDim - 1)>& global_corners_face1,
+    const std::array<size_t, two_to_the(VolumeDim - 1)>& global_corners_face2,
+    const std::vector<std::array<size_t, two_to_the(VolumeDim)>>&
+        corners_of_all_blocks) {
+  for (size_t j = 0; j < corners_of_all_blocks.size(); j++) {
+    const auto& corners_set = corners_of_all_blocks[j];
+    auto is_in_corners_set = [&corners_set](auto state, const auto element) {
+      return state and
+             std::find(corners_set.begin(), corners_set.end(), element) !=
+                 corners_set.end();
+    };
+    const bool face_matches_block_j =
+        std::accumulate(global_corners_face1.begin(),
+                        global_corners_face1.end(), true, is_in_corners_set)
+
+        and
+        std::accumulate(global_corners_face2.begin(),
+                        global_corners_face2.end(), true, is_in_corners_set);
+    if (face_matches_block_j) {
+      return j;
+    }
+  }
+  ERROR("This pair of faces does not belong to a single block.");
+}
+
+template <size_t VolumeDim>
+void DomainHelpers<VolumeDim>::set_internal_boundaries(
+    const std::vector<std::unique_ptr<EmbeddingMap<VolumeDim, VolumeDim>>>&
+        maps_of_all_blocks,
+    const std::vector<std::array<size_t, two_to_the(VolumeDim)>>&
+        corners_of_all_blocks,
+    std::vector<
+        std::unordered_map<Direction<VolumeDim>, BlockNeighbor<VolumeDim>>>&
+        neighbors_of_all_blocks) {
+  const double tolerance = 100 * std::numeric_limits<double>::epsilon();
+  for (size_t block1_index = 0; block1_index < corners_of_all_blocks.size();
+       block1_index++) {
+    std::unordered_map<Direction<VolumeDim>, BlockNeighbor<VolumeDim>> neighbor;
+    for (size_t block2_index = 0; block2_index < corners_of_all_blocks.size();
+         block2_index++) {
+      if (block1_index != block2_index and
+          DomainHelpers<VolumeDim>::get_common_global_corners(
+              corners_of_all_blocks[block1_index],
+              corners_of_all_blocks[block2_index])
+                  .size() == two_to_the(VolumeDim - 1)) {
+        ASSERT(DomainHelpers<VolumeDim>::physical_rms_separation(
+                   *(maps_of_all_blocks[block1_index]),
+                   *(maps_of_all_blocks[block2_index]),
+                   corners_of_all_blocks[block1_index],
+                   corners_of_all_blocks[block2_index]) < tolerance,
+               "Embeddingmaps "
```

Comment:
I put embedding map because the embedding map determines the physical location of the logical corners of the blocks